src/admin/__tests__/deleteTopicRecords.spec.js

```diff
@@ -0,0 +1,269 @@
+    brokerSpy.mockResolvedValueOnce() // succeed once
+    brokerSpy.mockRejectedValueOnce(new KafkaJSProtocolError('retriable error')) // fail once
+
+    await admin.deleteTopicRecords({ topic: topicName, partitions: recordsToDelete })
+
+    // broker call #1 succeeds, broker call #2 fails, call #3 should be the last one (skips broker #1, and only retries #2)
+    expect(brokerSpy).toHaveBeenCalledTimes(3)
+    expect(brokerSpy.mock.calls[1]).not.toEqual(brokerSpy.mock.calls[0])
+    expect(brokerSpy.mock.calls[2]).toEqual(brokerSpy.mock.calls[1])
+    brokerSpy.mockRestore()
+  })
+
+  test('in case offset is out of range, throws before processing any partitions', async () => {
+    recordsToDelete = [{ partition: 0, offset: '99' }]
+    const brokerSpy = jest.spyOn(Broker.prototype, 'deleteRecords')
+
+    await expect(
+      admin.deleteTopicRecords({ topic: topicName, partitions: recordsToDelete })
+    ).rejects.toThrow(
+      new KafkaJSProtocolError(
+        'The requested offset is not within the range of offsets maintained by the server'
+      )
+    )
+    expect(brokerSpy).not.toHaveBeenCalled()
+    brokerSpy.mockRestore()
+  })
+
+  test('in case offset is below low watermark, log a warning', async () => {
+    // delete #1 to set the low watermark to 5
+    recordsToDelete = [{ partition: 1, offset: '5' }]
+    await admin.deleteTopicRecords({ topic: topicName, partitions: recordsToDelete })
+    // delete #2
+    recordsToDelete = [
+      { partition: 0, offset: '7' }, // work as normal
+      { partition: 1, offset: '3' }, // logs a warning + no effect on the partition
+    ]
+    await admin.deleteTopicRecords({ topic: topicName, partitions: recordsToDelete })
+
+    expect(logger.warn).toHaveBeenCalledTimes(1)
+    expect(
+      logger.warn
+    ).toHaveBeenCalledWith(
+      'The requested offset is before the earliest offset maintained on the partition - no records will be deleted from this partition',
+      { topic: topicName, partition: 1, offset: '3' }
+    )
+    expect(
+      await cluster.fetchTopicsOffset([
+        {
+          topic: topicName,
+          partitions: [{ partition: 0 }, { partition: 1 }],
+          fromBeginning: true,
+        },
+      ])
+    ).toEqual([
+      {
+        topic: topicName,
+        partitions: expect.arrayContaining([
+          { partition: 0, offset: '7' },
+          { partition: 1, offset: '5' },
+        ]),
+      },
+    ])
+  })
+
+  test('in case partition does not exist/has no lead broker, skip the partition with a warning', async () => {
+    // try to add to the delete request a partition that doesn't exist
+    recordsToDelete = [
+      { partition: 0, offset: '7' },
+      { partition: 2, offset: '5' },
+    ]
+    await admin.deleteTopicRecords({ topic: topicName, partitions: recordsToDelete })
+
+    expect(logger.warn).toHaveBeenCalledTimes(1)
+    expect(logger.warn).toHaveBeenCalledWith('Could not find broker/leader for the partition', {
```

Comment:
I think the "broker/leader" wording is a bit confusing. Just the word "leader" is better, as we already use this terminology for the "LEADER_NOT_AVAILABLE" error: "There is no leader for this topic-partition as we are in the middle of a leadership election."